File: 1erdepartamental/c++/6.herencia_sobreescritura.cpp
// ============================================
// PROGRAMA: Herencia con Sobreescritura
// DESCRIPCIÓN: Implementaciones directa vs algorítmica
// ============================================

#include <iostream>
#include <stdexcept>
using namespace std;


/**
 * Base abstracta que obliga a implementar operaciones
 */
class CalculadoraBase {
protected:
    double __resultado;

public:
    CalculadoraBase() : __resultado(0) {}
    virtual ~CalculadoraBase() {}
    
    double getResultado() const { return __resultado; }
    
    // Métodos abstractos que DEBEN ser sobreescritos
    virtual double multiplicar(double a, int veces) = 0;
    virtual double potencia(double base, int exponente) = 0;
    virtual double dividir(double a, int divisor) = 0;
};


/**
 * Implementación directa usando operadores nativos
 */
class CalculadoraDirecta : public CalculadoraBase {
public:
    double multiplicar(double a, int veces) override {
        __resultado = a * veces;
        return __resultado;
    }
    
    double potencia(double base, int exponente) override {
        if (exponente < 0) {
            throw invalid_argument("Exponente debe ser >= 0");
        }
        __resultado = 1;
        for (int i = 0; i < exponente; i++) {
            __resultado *= base;
        }
        return __resultado;
    }
    
    double dividir(double a, int divisor) override {
        if (divisor == 0) {
            throw invalid_argument("División por cero");
        }
        __resultado = a / divisor;
        return __resultado;
    }
};


/**
 * Implementación algorítmica usando sumas/restas sucesivas
 */
class CalculadoraSucesiva : public CalculadoraBase {
public:
    // Multiplicación como sumas repetidas
    double multiplicar(double a, int veces) override {
        __resultado = 0;
        for (int i = 0; i < veces; i++) {
            __resultado += a;
        }
        return __resultado;
    }
    
    // Potencia como multiplicaciones repetidas
    double potencia(double base, int exponente) override {
        if (exponente < 0) {
            throw invalid_argument("Exponente debe ser >= 0");
        }
        __resultado = 1;
        for (int i = 0; i < exponente; i++) {
            __resultado *= base;
        }
        return __resultado;
    }
    
    // División como restas sucesivas
    double dividir(double a, int divisor) override {
        if (divisor == 0) {
            throw invalid_argument("División por cero");
        }
        __resultado = 0;
        double residuo = a;
        while (residuo >= divisor) {
            residuo -= divisor;
            __resultado++;
        }
        return __resultado;
    }
};


// ============================================
// EJECUCIÓN
// ============================================
int main() {
    cout << "=== CALCULADORA DIRECTA ===" << endl;
    CalculadoraDirecta cd;
    cout << "5 * 3 = " << cd.multiplicar(5, 3) << endl;
    cout << "2^8 = " << cd.potencia(2, 8) << endl;
    cout << "20 / 4 = " << cd.dividir(20, 4) << endl << endl;
    
    cout << "=== CALCULADORA SUCESIVA ===" << endl;
    CalculadoraSucesiva cs;
    cout << "5 * 3 = " << cs.multiplicar(5, 3) << endl;
    cout << "2^8 = " << cs.potencia(2, 8) << endl;
    cout << "20 / 4 = " << cs.dividir(20, 4) << endl;
    
    return 0;
}

File: 1erdepartamental/python/6.herencia_sobreescritura.py
from abc import ABC, abstractmethod
import cmath


class CalculadoraBase(ABC):
    def __init__(self):
        self.resultado = 0

    @abstractmethod
    def multiplicar(self, a, b):
        pass

    @abstractmethod
    def potencia(self, base, exp):
        pass

    @abstractmethod
    def dividir(self, a, b):
        pass


class CalculadoraDirecta(CalculadoraBase):
    def multiplicar(self, a, b):
        self.resultado = a * b
        return self.resultado

    def potencia(self, base, exp):
        self.resultado = cmath.exp(exp * cmath.log(base)) if base > 0 else base**exp
        return self.resultado

    def dividir(self, a, b):
        self.resultado = a / b if b != 0 else None
        return self.resultado


class CalculadoraSucesiva(CalculadoraBase):
    def multiplicar(self, a, b):
        self.resultado = 0
        for _ in range(abs(int(b))):
            self.resultado += abs(a)
        if (a < 0) != (b < 0):
            self.resultado = -self.resultado
        return self.resultado

    def potencia(self, base, exp):
        self.resultado = 1
        for _ in range(abs(int(exp))):
            temp = 0
            for _ in range(abs(int(base))):
                temp += abs(self.resultado)
            self.resultado = temp
        return self.resultado

    def dividir(self, a, b):
        if b == 0:
            return None
        cociente = 0
        residuo = abs(a)
        while residuo >= abs(b):
            residuo -= abs(b)
            cociente += 1
        self.resultado = cociente if (a >= 0) == (b >= 0) else -cociente
        return self.resultado, residuo


if __name__ == "__main__":
    cd = CalculadoraDirecta()
    print(f"Directa: 4*5 = {cd.multiplicar(4, 5)}")
    print(f"Directa: 2^4 = {cd.potencia(2, 4)}")

    cs = CalculadoraSucesiva()
    print(f"Sucesiva: 4*5 = {cs.multiplicar(4, 5)}")
    print(f"Sucesiva: 17/5 = {cs.dividir(17, 5)}")
